### ISSUE TITLE
bioframe 0.4.0 "select" incompatibilities

### ISSUE THREAD

**Issue by sergpolly** [open2c/cooltools#435]
tests are failing right now because of this #433 #434 and also because of some `bioframe.select` incompatibilities in 0.4.0
in a nutshell:
with bioframe 0.3.1 , cooler from tests `tests/data/CN.mm9.1000kb.cool` and track is just an EV from said cooler - same 2666 bins:
``` python
In [15]: clr.matrix().fetch(("chr1",100000000,150000000) ).shape
Out[15]: (50, 50)

In [16]: bioframe.select(track, ("chr1",100000000,150000000)).shape
Out[16]: (50, 4)
```
and with bioframe 0.4.0:
```python
In [4]: bioframe.select(track, ("chr1",100000000,150000000)).shape
Out[4]: (51, 4)
```

sample output of the bioframe.select here:
![image](https://user-images.githubusercontent.com/6790270/232927465-f9bf3ef6-8aaa-479c-979a-3d6cabb43f07.png)

that *99th* bins is not included with the older bioframe

**Comment by gfudenberg** [comment]
thanks for tracking this down!
Seems like this was addressed as Bug fix: treat query interval as closed on left
https://github.com/open2c/bioframe/commit/c22b6c20aeade98e82ca2b349a4b3d3651b22435
Which seems generally like the desired behavior... but I agree it is a bit weird for tracks.
Thoughts @nvictus  ?

**Comment by nvictus** [comment]
I think the "bug fix" was a mistake. Assuming (1) the query interval is to be interpreted as half-open, closed on the left and (2) the target intervals are also half-open, closed on the left, and (3) we want this operation to return any target interval that **overlaps** the query interval. Then, counterintuitively, the search query should look like:

```
mask = (query_chrom == df.chrom) & (query_start < df.end) & (query_end > df.start)
```

Typical case with bookended target intervals:
```
---
                  [QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ)
|      A       |        B        |      C      |       D        |         E         |
---
desired result: B, C, D
```

Edge-cases:
```
1. Query start position matches start of an interval
---
               [QQQQQQQQQQQQQQQQQQQQQQQQQ)
|      A       |        B        |      C      |       D        |
---
desired result: B, C (but not A)

2. Query end position matches end of an interval
---
       [QQQQQQQQQQQQQQQQQQQQQQQQQ)
|      A       |        B        |      C      |       D        |
---
desired result: A, B (but not C)

3. Special case of 2 and 3
---
               [QQQQQQQQQQQQQQQQQ)
|      A       |        B        |      C      |       D        |
---
desired result: B only
```

Can't remember if there were other edge cases where the behavior seemed less intuitive...

**Comment by sergpolly** [comment]
so it feel like this warrant a small release for cooltools then - with the versions of pandas and bioframe limited ?
maybe we can close a few maintenance PRs along with it

**Comment by gfudenberg** [comment]
we should definitely release a bioframe patch asap too.. I'm not sure how hard the pd issue is to fix, but we can easily roll back the select issue

**Comment by nvictus** [comment]
As I am fixing it, now I remember what the motivation was to change it. The additional edge case is if there are point intervals in the dataframe, the mask above will drop intervals with zero length, even if they lie inside the query interval.

**Comment by sergpolly** [comment]
So, in the end - bioframe is going to go back to half-open intervals and 0-length intervals are going to be "special-cased" - right ?

**Comment by nvictus** [comment]
Yes. The output should be consistent with the query and target intervals being half-open.

Using the mask formula above, the specific case that fails under that interpretation is when a 0-length interval lies exactly at `query_start` - it will be excluded from the selection.

This also applies to selecting from "point" dataframes, which is currently accomplished by using the same column as start and end, e.g. `cols=["pos", "pos"]`.

**Comment by nvictus** [comment]
Fixed with bioframe v0.4.1 release.

**Comment by nvictus** [comment]
@sergpolly go ahead and re-pin the dependency in cooltools
